File: OctoberChallenge/week1/rotate-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def rotateRight(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        if not head:
            return head;

        tmp = head;
        length = 1;
        while tmp.next:
            tmp = tmp.next;
            length += 1;

        if k > length:
            k %= length;
        k = length - k;
        if k==0 or k==length:
            return head;

        ## Handle current nodes:
        curr = head;
        count = 1;
        while count < k and curr:
            curr = curr.next;
            count += 1;

        if not curr:
            return head;

        ## Handle k-th node:
        nodeK = curr;
        tmp.next = head;
        head = nodeK.next;
        nodeK.next = None;
        return head;

File: OctoberChallenge/week1/rotate-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        // Handle head & base cases:
        if(! head) return head;

        ListNode* tmp = head;
        int len=1; //size of LinkedList
        while(tmp->next) { //TODO
            tmp = tmp->next;
            len++;
        }

        if(k > len) k %= len;
        k = len - k; //SOURCE OF ERROR
        if(k==0 || k==len) return head; //if nortation not needed

        // Handle current nodes:
        ListNode* curr = head;
        int count = 1;
        while(count < k && curr) {
            curr = curr->next; //update
            count++;
        }
        if(! curr) return head;

        //Handle k-th node:
        ListNode* nodeK = curr;
        tmp->next = head;
        head = nodeK->next;
        nodeK->next = NULL;
        return head; //not nodeK
    }
};
